<h2>1. Setup </h2>

<h3>1.1 Install Dependencies</h3>

In [1]:
pip install tensorflow==2.4.1 tensorflow-gpu==2.4.1 opencv-python matplotlib --ignore-installed --timeout 1000

  Using cached tensorflow-2.4.1-cp38-cp38-win_amd64.whl.metadata (2.7 kB)
  Using cached tensorflow_gpu-2.4.1-cp38-cp38-win_amd64.whl.metadata (2.7 kB)
  Using cached opencv_python-4.10.0.82-cp37-abi3-win_amd64.whl.metadata (20 kB)
  Using cached matplotlib-3.7.5-cp38-cp38-win_amd64.whl.metadata (5.8 kB)
  Using cached absl_py-0.15.0-py3-none-any.whl.metadata (2.4 kB)
  Using cached astunparse-1.6.3-py2.py3-none-any.whl.metadata (4.4 kB)
  Using cached flatbuffers-1.12-py2.py3-none-any.whl.metadata (872 bytes)
  Using cached google_pasta-0.2.0-py3-none-any.whl.metadata (814 bytes)
  Using cached h5py-2.10.0-cp38-cp38-win_amd64.whl.metadata (2.0 kB)
  Using cached Keras_Preprocessing-1.1.2-py2.py3-none-any.whl.metadata (1.9 kB)
  Using cached numpy-1.19.5-cp38-cp38-win_amd64.whl.metadata (2.0 kB)
  Using cached opt_einsum-3.3.0-py3-none-any.whl.metadata (6.5 kB)
  Using cached protobuf-5.27.1-cp38-cp38-win_amd64.whl.metadata (592 bytes)
  Using cached six-1.15.0-py2.py3-none-any.whl.met

<h3>1.2 Import Dependencies</h3>

In [2]:
# Import standard dependencies
import cv2
import os
import random
import numpy as np
from matplotlib import pyplot as plt

In [3]:
# Import tensorflow dependencies - Functional API
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Layer, Conv2D, Dense, MaxPooling2D, Input, Flatten
import tensorflow as tf


<h3>1.3 Set GPU Growth</h3>

In [4]:
# Avoid OOM errors by setting GPU Memory Consumption Growth
gpus = tf.config.experimental.list_physical_devices('GPU')
for gpu in gpus:
    tf.config.experimental.set_memory_growth(gpu, True)

<h3>1.4 Create Folder Structures</h3>

In [9]:
# Setup paths
POS_PATH = os.path.join('data', 'positive')
NEG_PATH = os.path.join('data', 'negative')
ANC_PATH = os.path.join('data','anchor')

In [14]:
# Make directories
os.makedirs(POS_PATH)
os.makedirs(NEG_PATH)
os.makedirs(ANC_PATH)

<h2>2. Collect Positives and Anchors</h2>

<h3>2.1 Untar Labelled Faces in the Wild Dataset</h3>

<h3>2.2 Collect Positive andAnchor Classes</h3>

<h2>3. Load and Preprocess Images</h2>

<h3>3.1 Get Image Directories</h3>

<h3>3.2 Preprocessing - Scale and Resize </h3>

<h3>3.3 Create Labelled Dataset </h3>

<h3>3.4 Build Train and Test Partition </h3>

<h2>4. Model Engineering </h2>

<h3>4.1 Build Embedding Layer </h3>

<h3>4.2 Build Distance Layer </h3>

<h3>4.3 Make Siamese Model </h3>

<h2>5. Training </h2>

<h3>5.1 Setup Loss and Optimizer </h3>

<h3>5.2 Establish Checkpoints</h3>

<h3>5.3 Build Train Step Function </h3>

<h3>5.4 Build Training Loop</h3>

<h3>5.5 Train the Model </h3>